### ISSUE TITLE
Type pattern is less efficient for value types than is/cast

### ISSUE THREAD

**Issue by jnm2** [dotnet/roslyn#28000]
If you use a type pattern, it creates an unnecessary local with an unnecessary load and store which are not optimized away.

```cs
return obj is SomeValueType someValueType && Equals(someValueType);
```

[Desugars](https://sharplab.io/#v2:EYLgHgbALAPgAgZgAQCcCmBDAJgewHYA2AnkgM4AuKArgMblIDKOAtmgGoYFVoAqRADmgCwAKADeopFKSIkOAG5oUKAJZY0SYDhwEkAUQCOVTqQAUOYACs0dOVYCUk6RJHS3MgOx3LSFaUYs7JzcfIJkgRxcvAIaAGSx+kYmpqQRwdGC9gDcTlIAvrlIhbJaOonGBGZMrJEhMXLkABZK9kgAvAB8SE0oOADuSHhUBAQ5IgUiQA==) to:

```cs
object obj2;
if ((obj2 = obj) is SomeValueType)
{
    SomeValueType other = (SomeValueType)obj2;
    return Equals(other);
}
return false;
```

It generates an extra four assembly instructions.

SharpLab asm: [type pattern](https://sharplab.io/#v2:EYLgHgbALANALiAhgZwLYB8ACBmABAJwFNEATAewDsAbAT12TnwFcBjOXAZTNUIDVEqTQgBUaAB0IBYAFABvGbkW4cuMgDdC+fAEsShXMDJkquAKIBHJgOQAKMsABWhNqscBKBUvnSlv5QHZXB1xtZE5uPgEhUQl6CP5BEXF9ADIUs0trG2R4qKSJNwBuT0UAXxLcCpVDYwyrKlsuHgTo5NU4AAtNN1wAXgA+XE78MgB3XAomKipi6XLpIA=) vs [`is`/cast](https://sharplab.io/#v2:EYLgHgbALANALiAhgZwLYB8ACBmABAJwFNEATAewDsAbAT12TnwFcBjOXAZTNUIDVEqTQgBUaAB0IBYAFABvGbkW4cuMgDdC+fAEsShXMDJkquAKIBHJgOQAKMsABWhNqscBKBUvnSlv5QHZXB1xtZE5uPgEhUQlcADI4s0trGxsuHn5BEXFCN3sHNwBuT0UAXxLcCpVDYySrKlt0yKyY/TI4AAtNN1wAXgA+XE78MgB3XAomKipi6XLpIA=)
SharpLab IL: [type pattern](https://sharplab.io/#v2:EYLgHgbALANALiAlgGwD4AEDMACATgUwEMATAewDtkBPbAZzlwFcBjObAZVIFt8A1Q5I3wAVKgAd8AWABQAbxnZF2LNlIA3fLlyJi+bMFKlk2AKIBHRgNoAKUsABW+VqocBKBUvnSlP5QHYXe2xEWg5uPgEhUQk6cP5BEXE9ADJk0wsra1o4yMSJVwBuD0UAX2LscpUDI3TLZBtOHniopNU4AAtNV2wAXgA+bA7cUgB3bHJGZGQi6TLpIA==) vs [`is`/cast](https://sharplab.io/#v2:EYLgHgbALANALiAlgGwD4AEDMACATgUwEMATAewDtkBPbAZzlwFcBjObAZVIFt8A1Q5I3wAVKgAd8AWABQAbxnZF2LNlIA3fLlyJi+bMFKlk2AKIBHRgNoAKUsABW+VqocBKBUvnSlP5QHYXe2xEWg5uPgEhUQlsADJY0wsra2tOHn5BEXF8Vzt7VwBuD0UAX2LscpUDI0TLZBs0iMzovVI4AAtNV2wAXgA+bA7cUgB3bHJGZGQi6TLpIA==)

/cc @gafter, @CyrusNajmabadi

**Comment by jnm2** [comment]
Is this intentional, on the off chance that someone took a ref to the parameter and is mutating it from a background thread? That scenario would be race-condition-prone either way—is that even worth considering?

**Comment by CyrusNajmabadi** [comment]
> That scenario would be race-condition-prone either way—is that even worth considering?

it's def worth considering given that hte language would not want you to be able to observe such a thing.  However, that said, we could consider some sort of optimization (or better codegen) when we can see that that local/parameter is never used in such a fashion.  In that case, the copy to the temp could be considered unnecessary and could optimize away.  @gafter for his thoughts on how viable any sort of approach is here.

**Comment by jnm2** [comment]
Wasn't thinking at first, but that scenario could result in InvalidCastException from the type pattern. Definitely bad. However, also not possible when the expression is a parameter or local. A ref to something on the stack can't be mutated concurrently with a type pattern's execution, right?

**Comment by Neme12** [comment]
I don't see what the concern here. Are race conditions well defined behavior in the language (as opposed to something like C++)?

**Comment by jnm2** [comment]
@Neme12 I would expect that `obj is Foo foo` should either fail or succeed but never throw InvalidCastException. I think it should protect against that. However, no need for that protection if `foo` is a local or parameter, for starters.

**Comment by Neme12** [comment]
I thought race conditions were always UB...

>  I would expect that obj is Foo foo should either fail or succeed but never throw InvalidCastException.

that's sounds to me like saying "I would expect `a++` to always fail or succeed" or "I would expect `a = b;` to always copy the all the data or nothing". If you have a race condition, InvalidCastException is probably one of the better outcomes.

**Comment by CyrusNajmabadi** [comment]
@Neme12 These are for locals/parameters.  He's saying: there should be no race conditions here.  So the compiler is being overly zealous in producing a temp.  it should avoid the temp and just do the work on the actual local/parameter as there's no way for other code to come in and affect the variable in an observable manner (including in multi-threaded environments).

**Comment by jnm2** [comment]
> If you have a race condition, InvalidCastException is probably one of the better outcomes.

I disagree. This is more similar to `?.Invoke` than anything else.

**Comment by Neme12** [comment]
> I disagree. This is more similar to ?.Invoke than anything else.

That is only the case for properties (which can really return anything at any time), not fields. If you're reading from a field and someone else is writing to it concurrently, you already have a problem.

My question is, why can't this optimization be done for fields as well?

**Comment by CyrusNajmabadi** [comment]
> My question is, why can't this optimization be done for fields as well?

@Neme12 It's unclear what you're referring to.  Do you mean not having the compiler write into a temporary for doing something like ```this.x is Z z```?

**Comment by Neme12** [comment]
yes

**Comment by CyrusNajmabadi** [comment]
> If you're reading from a field and someone else is writing to it concurrently, you already have a problem.

This is not true.  Lots of code can (and is) written with this working properly.  You have to be careful, but the language has defined semantics for what's going on.

Effectively, if you are doing anything defined as the language as multiple accesses, then all bets are off. So, for example, "a++" is a read and a write to a variable.  As those are two separate things, you have no guarantees about what may happen.

however, just accessing a variable is not undefined behavior in C#.  So code that just reads is safe.  And that means the compiler needs to properly generate IL that produces only a single actual read to prevent problems.

**Comment by Neme12** [comment]
> however, just accessing a variable is not undefined behavior in C#

How can that be true for arbitrarily sized data types?

**Comment by Neme12** [comment]
> So code that just reads is safe. 

Yes. I was talking about code that reads and writes at the same time.

**Comment by CyrusNajmabadi** [comment]
> Yes. I was talking about code that reads and writes at the same time.

There is no reason that has to be a problem.  And there is lots of code that will work even in the presence of that.  Again, you have to be careful, and know what you're doing.

**Comment by Neme12** [comment]
Ok sorry, I didn't know that C# guaranteed this.

**Comment by CyrusNajmabadi** [comment]
Also, calling this an optimization is a bit of a misnomer. it's normally a good thing to actually store this to a local. Consider if this is a field, if you have to read from it multiple times that might actually be quite expensive in practice. Storing into the local and just using that means avoiding any sort of memory lookups.

**Comment by CyrusNajmabadi** [comment]
The issue here is that storing into a temp (while good in general, and necessary for correctness in some circumstances) seems unnecessary when the original variable is a local or a parameter.

**Comment by jnm2** [comment]
@Neme12 One of the key value propositions of `?.Invoke` was that we'd no longer have to make a manual defensive copy in case someone unsubscribed from the event from another thread.

As further evidence, see the compare-and-swap loop the compiler generates every time you declare an auto event:
https://sharplab.io/#v2:EYLgZgpghgLgrgJwgZwLQAcEEsBusIAKCA9jBAMZkAmANDCFgDYA+AAgAwAErAjANwBYAFDDWAZm4AmTgGFhAb2Gdl3CRBwQAdjE4BRDdoASUTVUYQEnAMrEAthH1aYgoQF8gA==

**Comment by gafter** [comment]
This is fixed in the pending C# 8 implementation. See [HERE](https://sharplab.io/#v2:EYLgZgpghgLgrgJwgZwLRIMaOQSwG4SoAOsMECAdsgD4ACADAAS0CMA3ALABQ3yMCcDDEYBlAPYBbCADUoAGzgQAKgE8iEbgG9ujXcwDMjMQQQIcAEwiNgYsXMYBRAI5x5yABQB5YACsIQo18ASkZtLj0I5gB2QJ9GHGRRSRl5RVV1RmRk2QVlNSsAMgLHFzd3LKkctPygznC9AF8dPVpDGzsS1zkPcUrUvIyYEIBeAD5GfkU6hqA===).
